gadget/raw.go

```diff
@@ -0,0 +1,301 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package gadget
+
+import (
+	"bytes"
+	"crypto"
+	_ "crypto/sha1"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+// RawStructureWriter implements support for writing raw (bare) structures.
+type RawStructureWriter struct {
+	rootDir string
+	ps      *PositionedStructure
+}
+
+// NewRawStructureWriter returns a writer for the given structure, that will load
+// the structure content data from the provided root directory.
+func NewRawStructureWriter(rootDir string, ps *PositionedStructure) (*RawStructureWriter, error) {
+	if ps == nil {
+		return nil, fmt.Errorf("internal error: *PositionedStructure is nil")
+	}
+	if !ps.IsBare() {
+		return nil, fmt.Errorf("internal error: structure %s is not bare", ps)
+	}
+	if rootDir == "" {
+		return nil, fmt.Errorf("internal error: root directory cannot be unset")
```

Comment:
what is this directory in practice?  the mount dir of a gadget revision?  we have main things called root directory.

gadget root directory cannot be unset ?

Reply:
That's a mount directory of the gadget snap. Maybe `gadget content directory cannot be unset`?

Reply:
that sounds ok